### PR TITLE
IGNITE-20503 Sql. Support big clusters by mapping service

### PR DESCRIPTION
**Pull request by AMashenkov**
```
Benchmark                                    Mode  Cnt         Score        Error  Units
MappingBenchmark.benchBig                   thrpt   10   5990315.448 ±  45032.965  ops/s
MappingBenchmark.benchBigPartitionedOnly    thrpt   10  12656536.432 ±  47579.370  ops/s
MappingBenchmark.benchSmall                 thrpt   10  10608335.005 ±  22555.815  ops/s
MappingBenchmark.benchSmallPartitionedOnly  thrpt   10  20105426.118 ± 512649.606  ops/s
```

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/largecluster/AbstractTarget.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    abstract ExecutionTarget colocate(AllOfTarget other) throws ColocationMappingException;
+
+    abstract ExecutionTarget colocate(OneOfTarget other) throws ColocationMappingException;
+
+    abstract ExecutionTarget colocate(PartitionedTarget other) throws ColocationMappingException;
+
+    abstract ExecutionTarget colocate(SomeOfTarget other) throws ColocationMappingException;
+
+    static ExecutionTarget colocate(AllOfTarget allOf, AllOfTarget otherAllOf) throws ColocationMappingException {
+        if (!allOf.nodes.equals(otherAllOf.nodes) || otherAllOf.nodes.cardinality() == 0) {
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/largecluster/AbstractTarget.java:113]:
```suggestion
        if (!allOf.nodes.equals(otherAllOf.nodes) || otherAllOf.nodes.isEmpty()) {
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/largecluster/AbstractTarget.java [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+            throw new ColocationMappingException("Targets are not colocated");
+        }
+
+        if (!oneOf.nodes.get(target)) {
+            throw new ColocationMappingException("Targets are not colocated");
+        }
+
+        // When colocated, AllOfTarget must contains a single node that matches one of OneOfTarget nodes.
+        assert allOf.nodes.cardinality() == 1;
+        assert allOf.nodes.intersects(oneOf.nodes);
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/largecluster/AbstractTarget.java:133]: all these asserts looks like condom on a candle ?

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/ExecutionTargetFactorySelfTest.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+                new LargeClusterFactory(ALL_NODES)
+        );
+    }
+
+    @SuppressWarnings({"ResultOfObjectAllocationIgnored", "ThrowableNotThrown"})
+    @Test
+    void smallClusterFactory() {
+        List<String> nodes = IntStream.range(0, 65).mapToObj(i -> "node").collect(Collectors.toList());
+
+        assertThrows(IllegalArgumentException.class, () -> new SmallClusterFactory(nodes),"Supported up to 64 nodes");
```
Comment (zstan) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/ExecutionTargetFactorySelfTest.java:68]:
```suggestion
        assertThrows(IllegalArgumentException.class, () -> new SmallClusterFactory(nodes), "Supported up to 64 nodes");
```

**Review by zstan — APPROVED** [review]

**Review by korlov42 — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/smallcluster/AbstractTarget.java [hunk 4]
```diff
@@ -144,8 +156,17 @@ static ExecutionTarget colocate(OneOfTarget oneOf, PartitionedTarget partitioned
...
+        long colocatedNodes = oneOf.nodes;
+        for (int partNo = 0; partNo < partitioned.partitionsNodes.length; partNo++) {
+            colocatedNodes &= partitioned.partitionsNodes[partNo];
+
+            if (colocatedNodes == 0) {
+                throw new ColocationMappingException("Targets are not colocated");
+            }
         }
 
         return partitioned;
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/smallcluster/AbstractTarget.java:172]: you have to create new partitioned target with `colocatedNodes` for every partition. Here is an example: result of colocation of `PartitionedTarget(1=[A, B, C], 2=[C, D, E])` and `OneOfTarget(C, F)` is `PartitionedTarget(1=[C], 2=[C])`

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/largecluster/AbstractTarget.java [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        }
+
+        boolean changed = false;
+        BitSet newNodes = (BitSet) oneOf.nodes.clone();
+        for (int partNo = 0; partNo < partitioned.partitionsNodes.length; partNo++) {
+            if (!newNodes.equals(partitioned.partitionsNodes[partNo])) {
+                changed = true;
+            }
+
+            newNodes.and(partitioned.partitionsNodes[partNo]);
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/largecluster/AbstractTarget.java:167]:
```suggestion
            if (newNodes.equals(partitioned.partitionsNodes[partNo])) {
                continue;
            }
            
            changed = true;

            newNodes.and(partitioned.partitionsNodes[partNo]);
```

**Review by korlov42 — APPROVED** [review]

**Pull request merged**